### PR TITLE
Add an optional requirement for segment deletability

### PR DESCRIPTION
**Pull request by ncabatoff**
When the option is enabled, attempts to delete will block until all impacted segments are marked deletable.

### CODE REVIEW

File: wal.go [hunk 1]
```diff
@@ -858,7 +885,18 @@ func (w *WAL) truncateHeadLocked(newMin uint64) error {
 		return fin, postCommit, nil
 	})
 
-	return w.mutateStateLocked(txn)
+	for {
+		err := w.mutateStateLocked(txn)
+		switch err {
+		case nil:
+			return nil
+		case ErrUndeletable:
```
Comment (hghaf099) [inline wal.go:893]: should there be a deadline for this ? what if it gets stuck?

File: wal.go [hunk 2]
```diff
@@ -411,6 +415,20 @@ func (w *WAL) GetSealedLogFiles(fromIndex uint64) ([]*SealedSegmentInfo, error)
 	return sealedSegInfo, nil
 }
 
+func (w *WAL) MarkSealedLogDeletable(minIndex uint64) error {
```
Comment (hghaf099) [inline wal.go:418]: This seems that this only works if the minIndex (which is baseIndex) is given. Could there be a case where we don't know necessarily what is the minIndex/baseIndex and yet we want to delete a segment which contains Index `X` even if `X` is not the min/base Index? 
For that, we probably need to use `findSegmentReader` api. 
If we go with that we could avoid a another API call to get the base index and then try to mark it as deletable.

File: wal.go (hunk 1 above)
Comment (ncabatoff) [reply wal.go:893]: My thinking is, if the deletable stuff is enabled, it's the library user's (i.e. Vault's) job to ensure that it won't get stuck, by discovering sealed segments in a timely manner, and flagging them as deletable quickly.

File: wal.go (hunk 2 above)
Comment (ncabatoff) [reply wal.go:418]: I was assuming that this interface is to be used externally, feeding in information obtained by GetSealedLogFiles, which contain the baseIndex.

File: wal.go (hunk 1 above)
Comment (hghaf099) [reply wal.go:892]: Hmm, what do you mean by quickly?

File: wal.go (hunk 2 above)
Comment (hghaf099) [reply wal.go:417]: Ah, make sense!

**Review by hghaf099 — APPROVED** [review]
LGTM!

File: wal.go (hunk 1 above)
Comment (ncabatoff) [reply wal.go:892]: Within a few seconds.

**Review by raskchanky — APPROVED** [review]

**Pull request merged**